### ISSUE TITLE
Validating a process execution document without yet executing the process

### ISSUE THREAD

**Issue by jerstlouis** [opengeospatial/wps-rest-binding#101]
In the context of building a bridge between our server supporting the _Workflows_ extension with a core _OGC API - Processes_, we realized it would be very useful to validate the submission of a process execution document without actually triggering the processing.

This could be as simple as introducing a "mode" : "validate", and would allow the server to validate all inputs, and if successful return the list of outputs (without actual data) that would have been returned if the process was executed. This may also be related to the _Quotation_ extension (e.g. "mode" : "quotation").

In the proposed Workflows extension, a workflow invovling one or more processes can first be validated along the full chain to make sure it will work, before requests can be issued for specific AoI+resolution. Supporting this would allow to implement this bridge while properly validating processes from APIs not supporting the _Workflows_ extension, used as part of such a workflow.

**Comment by jerstlouis** [comment]
mode is gone now (#172)
The [Styles API](http://docs.opengeospatial.org/DRAFTS/20-009.html) has a `dry-run=true` query parameter.

**Comment by pvretano** [comment]
@jerstlouis so we can close this one?

**Comment by jerstlouis** [comment]
@pvretano no please don't close this one.

Maybe it could be tagged an extension, but ideally it should be included at least as a conformance class in Core to allow e.g. workflows to validate Core processes integrated as part of the workflows before actually doing any processing.

**Comment by bpross-52n** [comment]
It will remain an extension topic for now.

**Comment by bpross-52n** [comment]
SWG meeting from 2024-03-18: Add mechanism to validate a execution request without actually executing it. See Testbed-14 Quotation and Billing APIs for harmonization.

**Comment by fmigneault** [comment]
> validate Core processes integrated as part of the workflows before actually doing any processing.
> A mechanism to ask the server: will you be able to execute this request?

I would like to highlight that this is part of what the `POST /processes` DRU's role can accomplish when submitting the `executionUnit` consisting of a Workflow, to validate if the workflow is valid in order to execute it afterward. This ties in nicely with @pvretano's comment during the SWG meeting of 2024-07-08.

**Comment by jerstlouis** [comment]
@fmigneault Yes, DRU usually implies validation, but the difference between DRU and ad-hoc execution (which could validate or quote only) are:
- ad-hoc execution doesn't create a new process resource available to everyone accessing the API (extra privacy / security / authentication considerations),
- ad-hoc execution requires completely filled in inputs, whereas workflows targeting deployment as a process usually (though not always e.g. "ringing a bell process") map input parameters to the process to inputs within the workflow, and outputs of the workflow processes to outputs of the deployed process.

**Comment by fmigneault** [comment]
> extra privacy / security / authentication considerations

I think those are non issues. Our server handles all of them without problem. Only the user creating the process (or groups/users that he shared it with) see the process when listing `/processes`. 

> ad-hoc execution requires completely filled in inputs [...]

My main concern is that we keep duplicating endpoints and capabilities to support ad-hoc workflows, whereas everything would be completely transparent if those were handled as DRU, whether they are a simple process or a very convoluted workflow. The simplification of ad-hoc, while useful for simple use cases, seems to always over-complicate all other cases.

**Comment by pvretano** [comment]
I think I am agreeing with @fmigneault here.  Adding this capability just adds unnecessary complexity.  Even in the case of an ad-hoc workflow you expect the server to do some validation before executing anything and if that validation fails, the server would throw an exception.  So, say you have a workflow that is calling 3 processes.  I expect the server to verify that those other processes exist, that the inputs are valid according to the schema, etc. before even trying to execute the workflow.  If the workflow passes ALL the server's validation steps, only then does it get executed.  Am I missing something here?

**Comment by jerstlouis** [comment]
@pvretano 

>  If the workflow passes ALL the server's validation steps, only then does it get executed

The ability to do all this without launching an execution is extremely useful in the context of a workflow editor, for example.

It's also very useful in the context of the virtual _Collection Output_ execution mechanism, where you want to make sure the whole workflow works before returning a valid virtual collection. Such a workflow may include Part 1: Core process execution from servers that do not support Collection Output, which was the primary reason for me suggesting this.

**Comment by fmigneault** [comment]
@jerstlouis 

Under which context would you consider a Workflow as "built enough" to validate it, but not ready "final enough" to execute as is? If the workflow cannot run entirely, can it actually be considered valid?

In the case of Virtual Collection (if I understood its intended use correctly), wouldn't it still not be executed even if it was POST'ed, since it depends on incoming requests to trigger it from collection requests? Wouldn't the only way to actually validate the workflow for a Virtual Collection to actually submit it with a subset to validate it end-to-end?

**Comment by jerstlouis** [comment]
> Under which context would you consider a Workflow as "built enough" to validate it, but not ready "final enough" to execute as is?

@fmigneault It is not at all a matter of being "built/ready/final enough" or not, it is a matter of wanting to spend the computing resources at this point in time or not. A valid workflow should under most circumstances (99.99% of the time) be expected to complete successfully (the 0.01% being power outages, sharks eating underwater fibre optics, cosmic rays., large meteor strikes etc.).

> In the case of Virtual Collection (if I understood its intended use correctly), wouldn't it still not be executed even if it was POST'ed, since it depends on incoming requests to trigger it from collection requests? 

That is correct, but a virtual Collection Output may include as part of the workflow "Remote Core Processes", like the example we're discussing in T20-GDC where our Collection Output may rely on a Weaver process that we execute using Part 1 where your implementation does not yet support virtual Collection Output.

> Wouldn't the only way to actually validate the workflow for a Virtual Collection to actually submit it with a subset to validate it end-to-end?

If all component processes of the workflow support virtual Collection Output, then the nested workflow are essentially considered validated if each component returns a valid collection. Internally, each processing engine returning a valid collection should have done any necessary checks (including with any remote parties) to ensure that a request for subset/tile/dggs zone will succeed.

For a Part 1-only implementation, this "validate only" could be used for example as part of that validation process.
For example, before we return a virtual collection for our T20-GDC workflow integrating a Weaver process turning a STAC collection into GeoTIFF, we could submit to Weaver the nested portion of the workflow pointing to the STAC collection, and Weaver could let us know whether it will be able to successfully give us GeoTIFF for valid subsets.

We may need to include a small test subset as part of that Part 1 validation execution request if it's required by the process, but Weaver might be able to infer that this will likely (99.99%) work without actually doing any of the hard work.

**Comment by fmigneault** [comment]
I guess it comes down to each's definition of "validated". 
IMO, if the process didn't actually run, it is still not 100% validated (omitting the weird astral anomalies or sporadic network errors).

The workflow might be semantically correctly defined, all nested processes resolved, chaining steps compatible, etc., basically making it "seem valid", but something could still break because the server could apply additional restrictions that only occur when doing the actual execution request (eg: coverages refusing to respond to a too large subset) or it could mishandle the virtual collection trigger operations (since `response=validate` is used instead of `response=collection`, not the "entire" workflow is actually done). The processes can all give you "promises" that they would return some format or media-type, yet still return something different from expected behavior (either bug or partial/missing metadata). Many things could happen still leading to errors after only validation.

I would expect the most common use case is that the user believes their workflow is properly defined and submits it immediately. The execution engine would do the exact same validation operations than the proposed "validation mode", plus would also run the processing. If any error happens at any step ("pre-validation" or execution), the user has to debug their workflow anyway. If, somehow, the POST execution step passed without raising any easy-to-detect validations (aka: what "validation mode" would do), it would probably not be detected any earlier than at execution time anyway.

All in all, I believe this mode adds complexity to the implementation, with little to no actual extra validation of the processing chain than what should already be done anyway when submitting the execution request.

**Comment by jerstlouis** [comment]
> IMO, if the process didn't actually run, it is still not 100% validated 

> Many things could happen still leading to errors after only validation.

This is the essence of our disagreement.
I believe this validation to be possible assuming certified compliant implementations which passed all tests and a dry test run of the workflow that skips the bandwidth/computing-intensive aspects of the processing.

> eg: coverages refusing to respond to a too large subset

This kind of situation should be avoided by looking at server limits if the processing engine is acting as an OGC API coverage client from a `collection` Input, or by e.g., a HEAD request for a coverage request URL passed by `href`.

> I would expect the most common use case is that the user believes their workflow is properly defined and submits it immediately.

This is how OGC API - Processes is mostly used at the moment, yes.
But workflows could also be built / edited / shared without running them.

> If, somehow, the POST execution step passed without raising any easy-to-detect validations (aka: what "validation mode" would do), it would probably not be detected any earlier than at execution time anyway.

I really strongly disagree that you need to perform all processing before being able to fully validate a workflow.
The only reason why this may be the case is precisely because of the lack of such "validation mode" at each step of the workflow outside of one implementation's control, which is an argument for defining this validation-only capability.

> All in all, I believe this mode adds complexity to the implementation, with little to no actual extra validation of the processing chain than what should already be done anyway when submitting the execution request.

We're talking about a separate optional conformance class supporting a `response=validate` query parameter where the same execution request is submitted and a response is returned whether it is valid / invalid (with feedback as to what is wrong).
Since this validation is done anyways prior to and/or during execution, I don't see this as adding significant complexity.

I see this as extremely useful functionality. It is built into the virtual "Collection Output" mode (you only return a virtual Collection if the workflow is valid), but having this capability for sync/async execution would also be very useful (especially if you try to integrate Part 1-only processes into a workflow generating a Part 3 virtual collection).

This validation conformance class doesn't need to make it into this Part 1: Core 2.0 if there's not enough interest at this point (though I think that would have been the ideal place to include it), but I think we should at least consider it for a future extension.

In terms of being a dry run that doesn't actually perform the costly operations, it is very similar to the quotation mode, where the result is the expected execution time / cost as opposed to a valid/invalid response. If an implementation returns a quote for X hours and Y $, as a user if I then request authorization/budget for these costs, and then proceed to give the go ahead to run this, I expect to wait X hours and pay Y $ and get the results, and I would be upset to receive a notification 5 minutes later saying that the workflow failed for some obscure reason.

**Comment by jerstlouis** [comment]
By the wa I believe this validation / quotation / billing stuff has Part 2 / Application Package implications, since individual process would likely need a way to run the execution units in a validation / quotation dry-run mode, and for billing, execution should return billing details alongside the results.

**Comment by fmigneault** [comment]
> I really strongly disagree that you need to perform all processing before being able to fully validate a workflow.
The only reason why this may be the case is precisely because of the lack of such "validation mode" at each step of the workflow outside of one implementation's control, which is an argument for defining this validation-only capability.

A simple example. A LLM process interprets some generic query from a user about georeference entities, and returns the corresponding GeoJSON element it tries to guess information about. Then, a following process finds the relevant metadata from a database of georeferenced entities to extract desired information if possible. Since the entities to look for, their type, etc. are not known in advance, there is no way to pre-validate that the receiving database will accept them.  All processes might agree on their generic geo-formats end-to-end, but if the final search only accepts certain combinations of type/metadata, the workflow can still fail entirely.

The validation operation adds nothing but telling "it might work, just try it...".

I don't say the validation operation might not be useful in certain cases, but I expect these to be fairly limited "on its own" in comparison to simply trying the workflow and doing the validation inline with a reduced subset and get a better picture of the result, without any additional implementation requirement/complexity from the API and process definitions.

> quotation mode, where the result is the expected execution time / cost as opposed to a valid/invalid response

The quotation extension justifies a new endpoint because its purpose is much more than just estimating. It is also to generate a stored quote as a proof/receipt for the user which you can submit the process, manage as its own entity, and ensure agreed billing and assigned resources will be respected and persisted. It implies a legal contract. It "locks" the references, so the workflow cannot suddenly decide to resolve in a slightly different manner/location at a later stage. 

The estimates are supposed to be based on previous runs or any other logic you, as a server provider, are ready to assume if there is any issue or disparity from the agreed contract. An implementation is free to consider "validation" as good enough and throw random numbers for time/cost rather than basic its quote estimate on actual runtime evidence, but that is a choice the provider assumes.

> I believe this validation / quotation / billing stuff has Part 2 / Application Package implications, since individual process would likely need a way to run the execution units in a validation / quotation dry-run mode, and for billing, execution should return billing details alongside the results.

Indeed, they do. There is actually an entire evaluation @pvretano and I did about the quoting/billing evaluation, leveraging cost estimator algorithms, but it has not been made into the draft definition due to other priorities. Rough details are here: https://pavics-weaver.readthedocs.io/en/latest/quotation.html and implemented in Weaver.

**Comment by jerstlouis** [comment]
> A simple example. A LLM process interprets some generic query from a user about georeference entities,

I'm not sure I would qualify this as a simple example, but if the workflow essentially involves a discovery step generating a completely different and wide variety of potential inputs, it is sort of a metaworkflow. One way to handle such workflows would be to consider it valid, so that it should succeed rather than fail, but the output might be an empty set if it was not able to generate any useful output.

> The validation operation adds nothing but telling "it might work, just try it...".

Taking your example again, if the exact same workflow produces useful output by simply changing a natural language query string (or specifying a different bounding box, or a pair of geographic coordinates), then I argue that the validation operation should succeed and it does have value in saying "this is a valid workflow". Whether the output is an empty set or a useful result is a different story.

> I don't say the validation operation might not be useful in certain cases, but I expect these to be fairly limited "on its own" 

This extension feature request is about those certain cases, which I think are many.
The primary use case I was thinking of is trying to implement virtual Collection Output where an input is an OGC API - Processes - Part 1 server which does not itself support Collection Output.

You can indeed just "try it out" by trying a test execution request for a small subset -- this is what our _MOAWAdapter_ process/client currently tries to do. But this is significantly more cumbersome because:
- Assuming the process has something like an input parameter specifying the area of interest as a `bbox` data type and that the processing engine can identify that correctly, the success of this test depends entirely on a valid bbox being used for that test purpose, which is something our processing engine is ill-equipped to guess correctly
- By being foced to submit an execution request, there is no way for the other end to know we're just doing a test and to optimize out heavy bandwidth/calculations and all we care about is whether this request works or not and will produce the expected type of output 
- The "test" involves all of the complexity of executing a process such as polling if only async execution is supported

If we can perform this experiment with your Weaver implementation process accepting a STAC collection as an input and returning a GeoTIFF as an output, we will be using exactly this "test execution" just to validate the Weaver nested portion of the workflow works, and I can show all of the messy and lengthy steps involved which are needed before our implementation creates this virtual collection, which would be greatly streamlined/simplified if Weaver supported this `response=validate` mode.

> Rough details are here and implemented in Weaver.

Cool! Thanks!

**Comment by fmigneault** [comment]
> One way to handle such workflows would be to consider it valid, so that it should succeed rather than fail, but the output might be an empty set if it was not able to generate any useful output.

That somewhat goes against the point of the validation mode. It's like wrapping your entire code in a try/catch and ignoring all exceptions. Any error along the way are silently ignored, and the only way to know if it "really" works is still to execute it completely with a valid example.

>  if the exact same workflow produces useful output [...]

Exactly. It produces outputs, so it ran, even if just with a toy example or subset of data.
So, it still does not involve "only validation". It does the execution, which includes pre-validation.

> The primary use case I was thinking of is trying to implement virtual Collection Output [...]

I'm still failing to understand how the virtual collection output gets validated, since `response=validate`, the `response=collection` is not provided anymore. How does the engine get any hint of the intended virtual collection?

> which would be greatly streamlined/simplified if Weaver supported this `response=validate` mode.

I'm not sure what exactly this would accomplish beside parsing the already available target process description and just trusting the metadata it provides and OAP implementation behaves exactly as you expect, considering there are subtleties to support OAP v1/v2 and various combinations of optional requirements.

**Comment by jerstlouis** [comment]
> I'm not sure what exactly this would accomplish

I think if you could provide this example STAC Collection -> GeoTIFF output I could:
- A) try to get this TIE working so we can demonstrate a working STAC --> OGC APIs bridge
- B) try to illustrate the current challenges validating this without this `validate` mode, and how a `validate` mode would both simplify this validation process, speed it up (fewer HTTP requests / less processing), as well as provide a better guarantee of correctly asserting whether the Weaver-specific portion of the workflow is valid or not by removing guesswork on our side

Example scenario where validation of the nested workflow would fail:
- An invalid execution request
- Using capabilities in the execution requests not supported by the implementation (e.g., unsupported Part 3 capabilities)
- Using an OGC API Collection Input which only offers access mechanisms not supported by the implementation (e.g., OGC API - Coverages output supported by the collection, but processing engine only accepting STAC collections)
- Coverages supported by the collection and processing engine, but collection supporting GeoTIFF output whereas processing engine only accepts netCDF input
- Any issue with the STAC collection provided (e.g., the process requiring some STAC metadata that is not present to work)
- The processing engine not able to produce the requested output format for the process graph and specified inputs for whatever reason

Ideally, one could "leave out" any area of interest parameters in such request -- that could be done by making a `bbox` type input optional. In validate mode, this would mean _Would any request return something valid?_ rather than _process the whole thing_.

As a Processes client, our processing engine can try to put together a workflow and validate against the conformance declaration from the server, but by asking Weaver directly _Is this likely to work?_, we can remove a lot of the guesswork and be a lot more confident that this step of the workflow involving Weaver will work.

> How does the engine get any hint of the intended virtual collection?

If the above assertions all pass, then our server would instantiate a virtual collection based on the selected output from the Weaver process description, plus any additional processing applied to it higher up in the workflow, with the expectation that the nested process will indeed return a valid GeoTIFF for valid areas of interest.

The _geoDataClass_ concept linking to a logical schema describing e.g. bands and/or any other mechanisms to describe those bands and extent (e.g., JSON schema minimum/maximum for bbox fields?) in output description would also be useful to populate the description of that virtual collection (both the extent in `/collections/{collectionId}` and the fields in `/collections/{collectionId}/schema`).

This information could come from the process description directly and/or alternatively from the input themselves for something like our _PassThrough_ process simply returning/combining input fields, or other processes where extents in particular is the same as the input(s) (either a union or intersection for multiple inputs).

All this is much more streamlined when using Collection Input -> Process -> Collection Output, since you can rely on the availability of the collection description and collection schema to describe the inputs at both ends.

It's quite a bit of patchwork to try to get this working with "Remote Core Processes", but the "validate" mode could potentially go to some length to help with this. Something like a link to a generic sample output in the requested output format with the number of bands/fields and data type for each (or a sample feature collection with properties), but for a generic area not necessarily containing actual computation output, could also be useful. But at this point this is starting to sound like providing a collection description and logical schema :)

**Comment by pvretano** [comment]
TL;dr!  Guys, clearly this is not a settled issue.  It is not clear to me that is even qualifies as "core" functionality anyway so I am going to take it off the table for V2.0.  If, in the future, we settle on what "validation" means and have an API proposal to support this functionality, we can consider a corrigendum with a new conformance class.

**Comment by jerstlouis** [comment]
Thanks @pvretano , sounds good. It is not "Core" functionality -- it would be an optional requirement class and was already labeled with an _extension_ tag. I agree we should not delay the Part 1.0: Core v2.0 for this.
A new conformance class could potentially be added in a future corrigendum, or in a separate extension part.

**Comment by fmigneault** [comment]
@pvretano 

Sorry. You know how our discussions usually go 😅 
You can ignore for the sake of Core V2. This is definitely a long term extension thing. 

@jerstlouis 

> process requiring some STAC metadata that is not present to work

> this is starting to sound like providing a collection description and logical schema

This is exactly the feeling I'm getting. If the process description needs to replicate all the fine details that define the collection, what's the purpose of the collection? It's also very problematic when a process allows many variants. That doesn't scale well, and is really inconvenient for users to define processes. They will simply not bother filling all the examples and details requested to fulfill what "validation" mode requires. 

The only aspects I expect to be available for pre-validation are the properties already provided through the current process description, and I assume an OAP client to already be capable of parsing it and validating that they are compatible on both ends when trying to chain processes together. The server shouldn't have to approve that the client does a good job. 

Since it is the invoking client that is trying to define a workflow chain, I don't see why the server should preoccupy itself to consider all "higher level" considerations. The server should only worry about the process it advertises, and its I/O as an atomic step. That server then acts as a client when invoking a nested process, validating anything it needs to do, but the nested process only worries about its own atomic validations (so on and so forth recursively). IMO, this is already possible without involving a new validation mode.

**Comment by jerstlouis** [comment]
> If the process description needs to replicate all the fine details that define the collection, what's the purpose of the collection?

The purpose of _Collection Input_ & _Collection Output_ is to greatly facilitate integration of Data and Processing, whether they are offered locally, or on a remote OGC API implementation (distributed workflow).  Once instantiated, a virtual Collection Output is identical to a regular data collection, so retrieving data from by a client like GDAL is exactly the same. That in turn makes it possible to use a regular data collection or a virtual collection exactly the same way as a Collection Input. With the collection description & schema, it also facilitates knowing whether a particular output is suitable as an input.

Regardless of whether an implementation implements collection input/output or not, the _process description_ should still describe the process in such a way that given particular input(s), you can deduct exactly what to expect from the output. This is not only because as a Processes client you might want to know that, but also because in order to implement Collection Output, the processing engine ideally needs to be able to understand what the process is going to generate to describe the virtual collection.

> I assume an OAP client to already be capable of parsing it and validating that they are compatible on both ends when trying to chain processes together. The server shouldn't have to approve that the client does a good job.

> validating anything it needs to do, but the nested process only worries about its own atomic validations (so on and so forth recursively)

This recursive atomic validation is exactly the idea, but there may be several components involved. For example, let's say a user / workflow editor puts together a workflow involving both the GNOSIS and Weaver processing engines, plus a third party STAC API collection. This workflow is submitted to our processing end-point. The idea is that the GNOSIS implementation can rely on the Weaver processing engine to validate its nested part of the workflow, which would mean Weaver checking its immediate input (the STAC API collection), while GNOSIS validates the Weaver process output against the input to its process.

It may be theoretically possible for the GNOSIS implementation to verify the entire depth of the workflow by looking at all the collection inputs, the `/conformance` of every implementations, the description of all processes involved, but this could easily miss some validation details, where each implementation knows better whether a particular set of inputs will work for a particular process. If each process implementation does its own validation, it also allows workflows where each implementation does not necessarily fully understand everything within a nested workflow, such as a workflow that might mix openEO and OGC API - Processes.

> IMO, this is already possible without involving a new validation mode.

It is only possible by either:
- using "Collection Output", if the implementation supports it (again, the purpose of this validation mode is mainly for implementations **_not_** supporting Collection Output),
- doing a full-depth analysis of the entire workflow (e.g., checking ourselves the STAC API input to the Weaver process and inspecting the Weaver conformance to make sure that it supports STAC and probably tons of guesswork in terms by looking at the _LoadSTAC_ process trying to figure out whether this STAC API will actually work with that process; or if a nested process from another implementation is passed to Weaver, doing a full validation of that ourselves and so on -- no more reliance on recursive atomic validation), or
- by submitting a "test" execution request, which has its own challenges (such as picking a correct area subset which doesn't happen to be where there's no data which may result in a failure which is not indicative of a problem with the workflow)

**Comment by fmigneault** [comment]
What I meant by "*what's the purpose of the collection?*" was rhetoric. 
If the collection in itself cannot describe everything needed and that the process must redefine or extend the corresponding metadata, then collection's purpose is inappropriate or incomplete in the first place. We agree about all the facilitation a collection should provide. If we assume that all its description and schemas would be agreed between a process input and the desired collection passed to it, you shouldn't need the process to pre-validate it for you. You should be able to already do it by comparing the input and the collection from the calling client.

> the process description should still describe the process in such a way that given particular input(s), you can deduct exactly what to expect from the output

That is the main issue. Not all processes are deterministic. Conditional operations exist. Statistical deviations can happen. Randomness can be part of a process' normal execution. Sometimes, there is no "expected output". In such cases, static validation cannot be assumed. This is why I question the validity of this mode at all. Running a "test execution" would be what gives you the clearest idea of what you get back in return.

>  which would mean Weaver checking its immediate input (the STAC API collection), while GNOSIS validates the Weaver process output against the input to its process

Doesn't that mean that the Weaver nested process would be executed to obtain that output? So, why the `mode=validation`? It can just be a normal execution. Otherwise, how is this output obtained for GNOSIS to validate it? If not executing the Weaver process, how is the process supposed to know what is the "expected output", since you would provide a new input it has never seen to test it? It would have no base of comparison to know if the result matches what would be expected.


I am under the impression that the `mode=validation` only moves the guess-work from the calling workflow to the underlying nested process, but it still remains only guess-work. Not only that, the nested process does not have the full contextual knowledge of the workflow invoking it, so its guess-work task is even harder.

**Comment by fmigneault** [comment]
> test run would always be for a small sample of the data, which is not necessarily representative of the whole (e.g., no land elevation contours in the ocean)

I agree with you about this.  However, my impression is that if you get different results/errors that are affected/caused only when the full workflow is scaled up, regardless of using a validation mode or a real execution with small samples, you will actually observe the problem only once it is scaled up. 

A process offering a validation mode would still be unaware of that scaling up context. The problem is not somehow avoided. If you somehow would be able to anticipate this scaling issue prior to running it, such that it could be added to the implementation of the process to provide the validation mode, then that prior could be annotated in the process description, removing the need for the validation mode altogether. 

> What it would do is look at the input, and try to figure out what kind of output it would generate 

Weaver deals with deployed processes by CWL, which invokes a docker runtime with various scripts or utilities within them. In most cases, it cannot guess anything more than what the process description indicates. It relies on the I/O definitions specified by the user deploying it as the "source of truth". Therefore, its validation would be limited to the same criteria that want any OAP client should be able to do.

**Comment by fmigneault** [comment]
What I meant by "scaling up" is in response to the concern "*small sample of the data, which is not necessarily representative of the whole*". Basically, if edge cases cannot be identified from a few small-sample execute requests, I do not see how a validation mode would magically detect them any better. The unhandled cases would only arise when encountered during the entire processing context happens (when the actual problematic request is executed). If those edge cases were known in advance, then they could be defined as additional sample execute requests to attempt by the workflow engine trying to chain the relevant processes. The processes being chained should not have to care or try handling every possible situation in the universe you might throw at them. It is simply not feasible or realistic.

Therefore, if neither small test samples nor a validation mode can detect those edge cases in advance, I see little value in the extra validation mode considering all the extra problems and work it might put on OAP implementers. The OAP is already hard enough to implement entirely. It is preferable to avoid yet again more execution modes if they can be avoided (which I argue is possible by execution requests).

> If the deployed processes don't implement this "validate" mode, the processing engine could fallback to either the dumb validation that you say clients could do, or test runs.

This is another point I find problematic with the validation mode. If you cannot rely on it for *every process* to integrate in a workflow, you simply cannot rely on it at all. The workflow engine would still need to handle the cases manually where processes/servers do not provide the validation mode. If the workflow engine is implemented in a way to handle those "no-validation-provided" processes, then it could simply do it the same way for all processes assuming validation mode is never provided.

Also, a process could simply "implement" a *very dumb* validation mode by returning "200 OK" without doing any checks at all. The workflow engine would be misguided in a false sense of validation, would not be in any way more valid, and would still fail when running the unexpected input. You can never completely control or strong-arm an implementation to do "good/sufficient validation". Therefore, I argue the workflow engine shouldn't over-rely on them and do any checks it deems necessary by itself. 

> deployed processes (application packages) include a process description corresponding exactly to the OGC API - Processes description, because that seems to imply that all the processing engine does is schedule jobs and execute them and there's nothing more to it

It does way more than that. It's an abstraction layer that separates concerns. It manages everything that goes around and in-between the input request, the execution, staging of files, providing relevant metadata, apply inline I/O conversions as needed, etc. For example, if the processing engine wasn't there, every single deployed process would have to manually handle *Collection Inputs* and all intermediate operations invoking the various access mechanisms, even when the *Application Package* only cares about applying a specific operation such as computing NDVI on a GeoTiff. With the processing engine handling these concerns, the *Application Package* doesn't need to worry about them. I can easily extend the capabilities of the server as a whole without impacting the processes (as I did when adding *Collection Inputs* support). Whether a process is static *Part1:Core* or *Part2:Deploy* is completely transparent to you as a client invoking them, and it allows me as a server provider to extend the processing that can be offered without "opening the hood" of the server each time a new process must be added. The *OAP description* is simply one way to represent the *Application Package* to provide an interface to OAP clients.  I could provide their *openEO* equivalent interface at a later stage, and *Application Package* definitions would be completely agnostic of them.

> this makes it difficult for sharing tools / libraries / code which every process can leverage, without having to re-include it all in a docker for every single process deployed, meaning you may have 1000 copies of GDAL in 1000 dockers if you have a 1000 processes

This is not how containers work. Layers can be shared between image builds, and commands can be modified while reusing the same container runtime, so you don't have any duplication of the libraries and tools invoked. They are simply isolated under a virtual layer you can move around. You can make all the same tools/libraries available to all the processes the same way they would be bundled bare-metal on some machine, but you don't depend on a specific server configuration for them to run. That configuration is portable anywhere. You can wipe/instantiate/dispatch as desired and in a replicable fashion. If anything, this makes it much more sharable, since people and reuse the runtime you provide to build upon new tools.

**Comment by jerstlouis** [comment]
>  The unhandled cases would only arise when encountered during the entire processing context happens

The kind of processing in my mind here is of the more generic processing algorithms as provided by libraries like SAGA, and CQL2 field deriving expressions. If boundary conditions happen for a subset of the data, they could result in NODATA values for this part of the data, but overall it results in a valid datacube / virtual collection that can be partially generated over separate requests.

The point of the validation mode is not to detect edge cases, but whether the described workflow would generally work and can produce a valid virtual output collection. This is something much easier to detect with an explicit validation mode than test runs.

> If the workflow engine is implemented in a way to handle those "no-validation-provided" processes, then it could simply do it the same way for all processes assuming validation mode is never provided.

It could, but assuming the workflow engine's validation avoids a test run when a remote server or application package advertises support for a validation mode, it could make use of this instead, potentially in addition to its own validation.

> It does way more than that. It's an abstraction layer that separates concerns.
> With the processing engine handling these concerns, the Application Package doesn't need to worry about them.

That's great then. What I meant is then the process description inputs/outputs might be more flexible than the application package's inputs/outputs, and specific aspects of the processing engine might be relevant to the validation beyond the application package's process dsecription (which may or may not be reflected in the description at `/processes/{processId}`).

> This is not how containers work.

As you can tell containers are not my forte :) Great if there's a way to re-use things between deployed processes.

**Comment by jerstlouis** [comment]
@fmigneault In the [associated draft PR](https://github.com/opengeospatial/ogcapi-processes/pull/423#issuecomment-2214580050) [you said](https://github.com/opengeospatial/ogcapi-processes/pull/423#issuecomment-2214580050):

> IMO, we should focus on integrating the quotation and billing definitions that were already developed (though unpublished) with @pvretano over adding a new validation endpoint. **The validation does essentially the same thing** as requesting the quote using a pseudo-execution request, but quoting adds the additional estimate and ID for tracking of the original inputs submitted which can be used for the later execution without replicating the request/inputs again to actually run it.

As I mentioned earlier, it might make sense for "validation" to be a conformance class of that future extension part.
Instead or in addition to the quotation information, the validation might provide a "valid/invalid status" (including optional error information) and potentially also the logical schema of the response corresponding to the submitted execution request (based on the specific inputs provided).

For nested workflows, you would want both the validation and quotation capability to also cascade in a similar fashion, so that none of the nested remote servers actually produce the inputs but validate/quote their own parts of the workflow.

**Comment by fmigneault** [comment]
> What I meant is then the process description inputs/outputs might be more flexible than the application package's inputs/outputs

Definitely! It is not a 1:1 conversion. Typically, additional details are provided by Process Description that extend what the Application Package defines, or vice versa. For example, this [OGC Process Description metadata](https://github.com/crim-ca/weaver/blob/5a4805e3f29d99f1e0164c4bc46098d96959838f/tests/functional/application-packages/EchoBoundingBox/deploy.yml#L11-L22) extend the specific `ogc-bbox` format, which is only a "plain JSON file" in its corresponding [CWL Application Package](https://github.com/crim-ca/weaver/blob/5a4805e3f29d99f1e0164c4bc46098d96959838f/tests/functional/application-packages/EchoBoundingBox/echo_bbox.cwl#L8-L13) input.

Some parts of the validations are done by the processing engine, some are done by the invoked CWL, some can be done by the invoked tool, and some can be done by all of them. The idea being that we try to detect errors as early as possible, but some validations are too abstract at the processing engine level to be performed by it, and are instead dispatched to the app that knows more specifically what to do or issues to look for. This mixture of "validation steps/levels" also complexifies the implementation of a "validation mode".

>  openEO does define a validation end-point for user-defined process graphs

Yes, but look at the description. It illustrates what I am referring to: 

- >  Checks whether the process graph is schematically correct and the processes are supported by the back-end. It MUST also check the arguments against the schema, but checking whether the arguments are adequate in the context of data is OPTIONAL.
- > Errors that usually occur during processing MAY NOT get reported, e.g. if a referenced file is accessible at the time of execution.

So, basically, all it does is run the I/O values that are specified against the I/O schema, *if provided*. Since the Process Description already advertise the `schema`, I expect those to be already handled by the workflow engine that attempts building a functional chain *if it knows them*. However, if nothing is known in advance (no test samples), nothing can be validated beyond the schema definition. 

A similar validation to openEO's process-graph should also be done when deploying/executing a CWL or OAP-Part3 workflow. It doesn't blindly accept the workflow. It must be semantically correct. However, I expect this to be already done automatically and inline when performing the request, without need of another `/validation` endpoint. I'm not against having such an endpoint, but I argue it doesn't add much more than what is already done by the workflow engine unless actual samples are passed for the process to test against. Also, if runtime I/O values of nested processes does not match what they advertised, errors are still going to be raised even if the graph was pre-validated.

**Comment by jerstlouis** [comment]
> I argue it doesn't add much more than what is already done by the workflow engine unless actual samples are passed for the process to test against

The essence of my side of the argument is that a particular logical schema (e.g, as provided at `/collections/{collectionId}/schema`) for each "Data cube/Feature Collection style" input combined with specific numerical / textual literal value for any "parameter" style input (e.g., [ElevationContours](https://maps.gnosis.earth/ogcapi/processes/ElevationContours?f=json) `geometryType`) goes a long way and in the common use cases I am thinking of are fine substitutes for "actual samples".

Since some of the inputs may be interdependent, distributing this validation so that each remote server can validate its own process allows each server to consider those inter-dependencies with the process's built-in knowledge of how they work in terms of validating inputs as well as the resulting logical schema of the outputs.

Just like each server would need to quote its own part in a distributed workflow.

**Comment by fmigneault** [comment]
Ok. That makes sense. 

The last consideration I have regarding the original PR description (https://github.com/opengeospatial/ogcapi-processes/issues/101#issue-724794482) is the way to submit this validation. 

I do not think using `mode: validate` either as `?mode=validate` query parameter or within the JSON body is appropriate. It causes conflicting use with OAP V1 that could still make use of `mode`, and further overloads the already complex endpoint. IMO, it should be a dedicated endpoint (ie: `/processes/{id}/validation`) similar to the execution endpoint, the proposed quotation endpoint, and what openEO does.

**Comment by pvretano** [comment]
Wouldn't the validation point need to be independent of a particular process ID?  If the thing being validated is a workflow it could conceivably reference many processes so `/processes/{id}/validation` does not seem appropriate.

**Comment by jerstlouis** [comment]
@fmigneault 

> Ok. That makes sense.

So happy to hear you say it -- maybe I'm finally starting to explain the Part 3 ideas properly :)
I hope I didn't just tire you out with my long arguments :P

@pvretano 

> Wouldn't the validation point need to be independent of a particular process ID? If the thing being validated is a workflow it could conceivably reference many processes so /processes/{id}/validation does not seem appropriate.

I don't really mind _how_ this validation thing can be done.  I was originally suggesting `.../execution?response=validate`.

We have not yet really settled on how ad-hoc nested processes workflows should work in Part 3.
As far as the submission of ad-hoc nested workflows defined using Part 1 execution requests, the current draft Part 3 "Nested/Remote Processes" still uses the same end-point as Part at `/processes/{processId}/execution`, because there really is a top-level process which will receive inputs from the others. So this is really the most natural way of extending Part 1, since you're just defining processes as input to the parent process.

For CWL and openEO, it may be a valid argument that there's no top-level process, and that posting to `/jobs` for sync/async as proposed in Part 4, and POST to `/collections` for virtual Collection Output (instead of `?response=collection`) make sense as alternatives.

For validation, it would make sense to have something that mirrors the quotation end-point, because I think it's very similar.

**Comment by fmigneault** [comment]
@jerstlouis 
No. I see the relevance of this operation if it is limited to the cases described in https://github.com/opengeospatial/ogcapi-processes/issues/101#issuecomment-2613496390, for additional cross-input validation only the process might be aware of, or for extra validation based on its internal definition that is not exposed "outside" via OAP Process Description. 

Where I find it becomes a stretch is when a given process has to somehow consider other processes in a nested workflow chain. This is why I believe @pvretano's worry about "independent of a particular process ID" *should* be a non-issue, because I believe a validation mode should be a process-atomic operation. If workflows are involved, it becomes the job of the client creating that workflow to handle the various validations involved by each process, and any additional validations involved in the "chain". 

At best, if a validation should be provided for a workflow, I believe this could be handled using openEO's approach (and now possible with OAP-Part4 Job Management proposal), to create that job as `status: created` pending execution, and then invoke some `/jobs/{jobId}/validation` endpoint.
